File: 프로그래머스/lv2/87946. 피로도/피로도.py
import itertools


def solution(k, dungeons):
    answer = 0
    cnt = 0

    order_list = list(itertools.permutations(dungeons))

    for fatigue in order_list:
        answer = max(answer, cnt)
        cur = k
        cnt = 0
        for i in range(len(fatigue)):
            if cur >= fatigue[i][0]:
                cnt += 1
                cur -= fatigue[i][1]
            else:
                break
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int d_len;
    static boolean[] visit;
    static int answer = 0;
    
    public int solution(int k, int[][] dungeons) {
        d_len = dungeons.length;
        visit = new boolean[d_len];
        dfs(k, 0, dungeons);
        return answer;
    }
    public void dfs(int hp, int depth, int[][] dungeons){
        
        if(depth == d_len) return;
        
        for(int i=0; i < d_len; i++){
            if(!visit[i] && hp >= dungeons[i][0]){
                visit[i] = true;
                answer = Math.max(answer, depth+1);
                dfs(hp - dungeons[i][1], depth+1, dungeons);
                visit[i] = false;
            }
        }
    }
}